subscriptions/providers/apple_in_app/__init__.py

```diff
@@ -211,35 +252,24 @@ def _handle_receipt(self, request: Request, payload: AppleReceiptRequest) -> Res
     def _handle_renewal(self, notification: AppStoreNotification) -> None:
         transaction_info = notification.transaction_info
 
-        # Check if we handled this before.
-        with suppress(SubscriptionPayment.DoesNotExist):
-            SubscriptionPayment.objects.get(
-                provider_codename=self.codename,
-                provider_transaction_id=transaction_info.transaction_id,
-            )
-            # If we didn't raise, it means that we've already handled this transaction, just the App Store didn't
-            # receive the information that we did. Skip it. If it wasn't renewal, we could start searching.
-            return
-
         latest_payment = self._get_latest_transaction(transaction_info.original_transaction_id)
-
-        # Information about this should've been received earlier via change, but downgrades are to happen
-        # only at the start of new cycle. So we handle them here.
-        current_product_id = latest_payment.plan.metadata.get(self.codename)
-        if current_product_id != transaction_info.product_id:
-            latest_payment.plan = self._get_plan_for_product_id(transaction_info.product_id)
-            # Create a new subscription.
-            latest_payment.subscription = None
-
-        latest_payment.pk = None
-        # Updating relevant fields.
-        latest_payment.provider_transaction_id = transaction_info.transaction_id
-        latest_payment.subscription_start = transaction_info.purchase_date
-        latest_payment.subscription_end = transaction_info.expires_date
-        latest_payment.created = None
-        latest_payment.updated = None
-        # Metadata stays the same.
-        latest_payment.save()
+        assert latest_payment, f'Renewal received for {transaction_info=} where no payments exist.'
+
+        current_plan = latest_payment.plan
+        subscription = latest_payment.subscription
+        if current_plan.metadata.get(self.codename) != transaction_info.product_id:
+            current_plan = self._get_plan_for_product_id(transaction_info.product_id)
+            subscription = None
+
+        self._get_or_create_payment(
+            transaction_info.transaction_id,
+            transaction_info.original_transaction_id,
+            latest_payment.user,
+            current_plan,
+            transaction_info.purchase_date,
+            transaction_info.expires_date,
+            subscription=subscription,
+        )
```

Comment:
If plan changed, you need to stop previous subscription and create a new one. Here I see only creation of new one without stopping of the old one. User will get resources for both. Why at all current plan may be different that transaction's product if it's just a renewal?

Also, new subscription uses `transaction_info.purchase_date`, are you sure it's the date when subscription was renewed? Or maybe you will have a payment which starts from the beginning of first subscription?

Reply:
I didn't stop the previous subscription since I assumed that this is called either on renewal on from an upgrade. I should do it explicitly here, would be easier to understand. Will fix.

As for the renewal with product change – Apple states that there are two events that can change the product. They're defined as an upgrade and downgrade. While upgrade is to take place immediately, downgrade is to happen at the next renewal.
https://developer.apple.com/documentation/appstoreservernotifications/subtype
```
(...) Downgrades take effect at the next renewal. (...)
(...) Upgrades take effect immediately. (...)
```
So, not to overcomplicate things, upgrade is performed immediately and downgrades are disregarded until the next renewal.

The `purchase_date`: https://developer.apple.com/documentation/appstoreservernotifications/purchasedate
```
The time that the App Store charged the user’s account for a purchase, a restored product, a subscription, or a subscription renewal after a lapse.
```
There's another field, `originalPurchaseDate`, for the start of the first subscription.

Reply:
Assigned `timezone.now()` to `end` of a subscription (and saved) before reseting it to `None` (and thus creating a new one).